### ISSUE TITLE
Custom Migration Template

### ISSUE THREAD

**Issue by cullylarson** [robmorgan/phinx#602]
There are a few things I always find myself changing from the default migration template, when creating a new migration (e.g. adding an init function to get the table prefix, changing comments, putting starting braces on the same line as the function def, etc.).  It would be helpful to be able to define a custom template, but currently it looks like the default migration template is hardwired.  I wouldn't mind working on an implementation, if it sounds like a good idea.

**Comment by cullylarson** [comment]
For some reason I didn't notice before, but I see that you can set a custom template via a command line argument (I just updated to version 4.4, maybe it's new?).  Anyway, it seems like it wouldn't be difficult to look for the alt template in the config, as well as from the command line (e.g. if not provided via command line, check config).

**Comment by IngwiePhoenix** [comment]
The „distribution provided“ template is at: vendor/robmorgan/phinx/src/Phinx/Migration/Migration.template.php.dist
To use your own template: $ phinx —template path/to/template.php

You can not set a template via the configuration, yet. Also, the template option is only relevant in „create“.

**Comment by rquadling** [comment]
Duplicate : https://github.com/robmorgan/phinx/issues/528

**Comment by laurin1** [comment]
I had to use -t, when I use -template, I get this:

  [InvalidArgumentException]
  The alternative template file "emplateC:\inetpub\Intranet_Local\bin\/../lib
  rary/phinx_migration_template.txt" does not exist

**Comment by laurin1** [comment]
So leaving the -t in works fine for migration and for create (I assume it ignores it in migration), but fails when I run test! That means if I am using a batch file, I have to now have 2 batch files which is a pain.

**Comment by rquadling** [comment]
The parameter has a long ```--template``` and a short ```-t``` format.

**Comment by rquadling** [comment]
The ```--template``` and ```--class``` parameters are only applicable to the migration creation command as they create the migration file that you would now edit and execute by using ```phinx migrate```.

As an example, this is one of our template creation classes ...

```
<?php
namespace xxx\Migration\Phinx;

class Permission extends PhinxCreation
{
    /**
     * Post Migration Creation.
     *
     * Once the migration file has been created, this method will be called, allowing any additional
     * processing, specific to the template to be performed.
     *
     * @param string $migrationFilename The name of the newly created migration.
     * @param string $className         The class name.
     * @param string $baseClassName     The name of the base class.
     *
     * @return void
     */
    public function postMigrationCreation($migrationFilename, $className, $baseClassName)
    {
        echo <<< END_MIGRATION

  The Permission migration file has been created.

  You only need to edit 3 elements in {$migrationFilename} and you are done!


END_MIGRATION;
    }
}
```

and the corresponding template file is ...

```
<?php

use xxxx\Migration\Permission;

/**
 * Class $className
 *
 * A xxxx Permission Migration.
 */
class $className extends Permission
{
    public function setup()
    {
        /**
         * The friendly name for this permission.
         *
         * Something like :
         * Reports / Daily Sales Analysis
         */
        $this->friendlyName = '';

        /**
         * The permission name.
         *
         * Something like:
         * report_daily_sales_analysis
         */
        $this->name = '';

        /**
         * The general permissions.
         *
         * This is a bitmask based upon the Permissions values of:
         * PERM_ADD - Allow Add permission
         * PERM_ALL - Allow All permissions
         * PERM_DELETE - Allow Delete permission
         * PERM_EDIT - Allow Edit permission
         * PERM_LIST - Allow List permission
         */
        $this->permissions = self::PERM_LIST;
    }
}
```

and, finally, the migration class called by this template (NOTE: we only use a single db and have been a bit lazy by using raw SQL), 

```
<?php
namespace xxxx\Migration;

use RuntimeException;

/**
 * Class Permission
 *
 * Migration for a permission.
 *
 * @package xxxx\Migration
 */
abstract class Permission extends AbstractMigration
{
    /**
     * Allow Add permission.
     */
    const PERM_ADD = 1;

    /**
     * Allow All permissions.
     */
    const PERM_ALL = 15;

    /**
     * Allow Delete permission.
     */
    const PERM_DELETE = 2;

    /**
     * Allow Edit permission.
     */
    const PERM_EDIT = 4;

    /**
     * Allow List permission.
     */
    const PERM_LIST = 8;

    /**
     * The friendly name of the permission as viewed in the permissions editing.
     *
     * Often displayed as a "breadcrumb" style name for reports, e.g. 'Report > Daily Sales'.
     *
     * @var string
     */
    protected $friendlyName = null;

    /**
     * The internal name of the permission.
     *
     * This is what is referenced in the code to limit functionality by permission.
     *
     * These are 'snake_cased'.
     *
     * @var string
     */
    protected $name = null;

    /**
     * The permissions that can be granted for this entry.
     *
     * This is a bitmask of PERM_xxxx values.
     *
     * @var integer
     */
    protected $permissions = null;

    /**
     * Rollback
     */
    final public function down()
    {
        $this->validate();

        if ($tableID = $this->getTableID()) {
            $this->writeComment(
                'Removing permission',
                trim(
                    ($this->permissions & self::PERM_ADD ? 'ADD ' : '') .
                    ($this->permissions & self::PERM_DELETE ? 'DELETE ' : '') .
                    ($this->permissions & self::PERM_EDIT ? 'EDIT ' : '') .
                    ($this->permissions & self::PERM_LIST ? 'LIST ' : '')
                )
            );
            $this->execute("DELETE FROM `xpermissions` WHERE `tableID` = {$tableID}");
        }

        $this->writeComment('Removing table     ', $this->friendlyName);
        $this->execute("DELETE FROM `tables` WHERE `name` = '{$this->name}' LIMIT 1");
    }

    /**
     * Migrate
     */
    final public function up()
    {
        $this->validate();

        $this->writeComment('Adding table     ', $this->friendlyName);
        $this->execute("INSERT INTO `tables` (name, friendlyName) VALUES('{$this->name}', '{$this->friendlyName}');");

        if ($tableID = $this->getTableID()) {
            $this->writeComment(
                'Adding permission',
                trim(
                    ($this->permissions & self::PERM_ADD ? 'ADD ' : '') .
                    ($this->permissions & self::PERM_DELETE ? 'DELETE ' : '') .
                    ($this->permissions & self::PERM_EDIT ? 'EDIT ' : '') .
                    ($this->permissions & self::PERM_LIST ? 'LIST ' : '')
                )
            );
            $this->execute(
                "INSERT INTO `xpermissions` (`usergroupID`, `tableID`, `add`, `delete`, `edit`, `list`) " .
                "VALUES(1, {$tableID}, " .
                ($this->permissions & self::PERM_ADD) . ", " .
                ($this->permissions & self::PERM_DELETE) . ", " .
                ($this->permissions & self::PERM_EDIT) . ", " .
                ($this->permissions & self::PERM_LIST) . ")"
            );
        }
    }

    /**
     * Prepare the permission migration.
     */
    abstract protected function setup();

    /**
     * Validate the permission setup.
     *
     * @throws RuntimeException
     */
    protected function validate()
    {
        $this->setup();

        if (empty($this->friendlyName)) {
            $this->writeError(' E R R O R', '$friendlyName cannot be empty');
            throw new RuntimeException('$friendlyName is empty');
        }
        if (empty($this->name)) {
            $this->writeError(' E R R O R', '$name cannot be empty');
            throw new RuntimeException('$name is empty');
        }
        if (empty($this->permissions)) {
            $this->writeError(' E R R O R', '$permissions cannot be empty');
            throw new RuntimeException('$permissions is empty');
        }
    }

    /**
     * Get the unique ID of the recently added permission.
     *
     * @return int
     */
    private function getTableID()
    {
        return (int)($this->fetchRow("SELECT `tableID` FROM `tables` WHERE `name` = '{$this->name}'")['tableID']);
    }
}
```

We also have an extended ```AbstractMigration``` class which allows us some additional formatted reporting and exception handling.

By using an alternative migration template (or a migration template generating class), you can build migrations that are fit your needs more than Phinx currently provides, but without losing all the benefits of Phinx (rollback, database abstraction, different environments, etc.).

**Comment by daison12006013** [comment]
You could mock the command line to your own command line, you can check my repo: https://github.com/phalconslayer/slayer/blob/master/bootstrap/src/Support/Phinx/Console/Command/Create.php

Wherein I extended Phinx\Console\Command\Create class and changing the Symfony command name, wherein you could also override the default Options ``--t`` to something.

**Comment by robmorgan** [comment]
closing as I merged 100f0188b1cf5342bb3bbc402ea88c859e5dda3b
